File: Structural/Proxy/src/examplejava/medialibrary/YouTubeCacheProxy.java
package examplejava.medialibrary;

import java.util.HashMap;

public class YouTubeCacheProxy implements YouTubeLibrary {
    private YouTubeLibrary youTubeService;
    private HashMap<String, Video> cachePopular = new HashMap();
    private HashMap<String, Video> cacheAll = new HashMap();

    public YouTubeCacheProxy() {
        this.youTubeService = new YouTubeClass();
    }

    @Override
    public HashMap<String, Video> popularVideos() {
        if (cachePopular.isEmpty()) {
            cachePopular = youTubeService.popularVideos();
        } else {
            System.out.println("Retrieved list from cache.");
        }
        return cachePopular;
    }

    @Override
    public Video getVideo(String videoId) {
        Video video = cacheAll.get(videoId);
        if (video == null) {
            video = youTubeService.getVideo(videoId);
            cacheAll.put(videoId, video);
        } else {
            System.out.println("Retrieved video '" + videoId + "' from cache.");
        }
        return video;
    }

    public void reset() {
        cacheAll.clear();
        cachePopular.clear();
    }
}

File: Structural/Proxy/src/examplekotlin/medialibrary/YouTubeCacheProxy.kt
package examplekotlin.medialibrary

import java.util.*

class YouTubeCacheProxy : YouTubeLibrary {
    private val youTubeLibrary: YouTubeLibrary = YouTubeClass()
    private var cachePopularVideos = HashMap<String, Video>()
    private val cacheAll = HashMap<String, Video>()

    override fun popularVideos(): HashMap<String, Video>{
        if (cachePopularVideos.isEmpty()){
            cachePopularVideos = youTubeLibrary.popularVideos()!!
        }
        else    {
            println("Retrieved list from cache")
        }
        return cachePopularVideos
    }

    override fun getVideo(videoId: String): Video {
        var video: Video? = cacheAll[videoId]
        if (video == null) {
            video = youTubeLibrary.getVideo(videoId)
            cacheAll[videoId] = video
        } else {
            println("Retrieved video '$videoId' from cache.")
        }
        return video
    }


}